CARDS-1623-1724: Add support for DICOM images through the cards:DicomAnswer question type

IntegralProgrammer opened the pull request:
This PR adds the `cards:DicomAnswer` question type which allows for questionnaires to accept DICOM medical imaging file uploads. Upon DICOM upload, the image is displayed in the Form and the DICOM's metadata is written to the _Notes_ field associated with the question. DICOM file parsing is done on the front-end using the Cornerstone.js library.

It is important to note that at this time, DICOM images that use a _TransferSyntaxUID_ of `1.2.840.10008.1.2.4.70` (JPEG Lossless, Non-hierarchical, First-Order Prediction - Processes 14) are unable to generate an image preview. Metadata extraction from these files still works nonetheless. Using the _dynamic-import_ version of Cornerstone, these previews could be generated, however integrating the _dynamic-import_ version of Cornerstone with CARDS would require upstream modifications to Cornerstone and therefore, for the sake of simplicity, the regular version of the Cornerstone library is used with external error handling for DICOM files with unsupported _TransferSyntaxUID_ values. When these fixes are made to the upstream Cornerstone library, the previously-unsupported _TransferSyntaxUID_ values can simply be removed from the `FORBIDDEN_TRANSFER_SYNTAX_UID` list.

**To test:**

- Build this (`CARDS-1623_1724`) branch with `mvn clean install`
- Start CARDS with the `test` features enabled so that the _DICOM Imaging Test_ Questionnaire is available (`./start_cards.sh --dev --test`).
- Visit `http://localhost:8080` and login as `admin`:`admin`.
- For the sample DICOM files on https://barre.dev/medical/samples/, create a _DICOM Imaging Test_ Form and upload a DICOM file.
- Some of these files (for example, `CT-MONO2-8-abdo.dcm`) will work without modification. That is, an image preview will be displayed and the _Notes_ file will be auto-populated with the DICOM file's metadata.
- Other files (for example `CR-MONO1-10-chest.dcm`) will generate an error message and suggest using `gdcmconv` to fix the file. Follow the suggestion and the converted file upload should work properly.
- The file `CT-MONO2-16-chest.dcm` uses a _TransferSyntaxUID_ of `1.2.840.10008.1.2.4.70`. Uploading this file will display its metadata in the _Notes_ field, but the image preview will be replaced with a message of `Failed to render preview due to Invalid TransferSyntaxUID` as the _TransferSyntaxUID_ (pixel encoding format) is not supported due to upstream issues in the Cornerstone library (see above).

**Regression testing:**

- `FileQuestion` and `Note` have been modified for the implementation of `DicomQuestion`. Test that there are no regressions in these two components.

marta- committed:
CARDS-1623: Add basic support for DICOM image storage
* DicomAnswer node type definition
* DicomQuestion component based on FileQuestion (not functional outside file upload)
* dicom question configuration for the questionnaire wizard
* Basic questionnaire containing a dicom question
* Minor style change to preserve whitespace and indentation in notes
(work in progress)

IntegralProgrammer committed:
CARDS-1623: Add basic support for DICOM image storage

Bring in the Cornerstone.js DICOM parser library

IntegralProgrammer committed:
CARDS-1623: Add basic support for DICOM image storage

Add a DICOM tag data dictionary taken from the Cornerstone.js git repository

IntegralProgrammer committed:
CARDS-1623: Add basic support for DICOM image storage
WIP - DICOM metadata is written to Notes field but Notes field is not saved to the back-end upon Form submission

marta- committed:
CARDS-1623: Add basic support for DICOM image storage
Added a label processor to expor the dicom file name as the `displayedValue` for dicom type questions

marta- committed:
CARDS-1623: Add basic support for DICOM image storage
Fixed saving the dicom metadata in the note

marta- committed:
CARDS-1623: Add basic support for DICOM image storage
Small refactoring/simplification of the code generating the dicom metadata note

marta- committed:
CARDS-1623: Add basic support for DICOM image storage
Styling adjustments for the DicomQuestion. This commit includes displaying temporarily placeholder image in lieu of the actual dicom image, for demonstration purposes.

marta- committed:
CARDS-1623: Add basic support for DICOM image storage
Expanded the customization capabilities of the Note component to better adjust to displaying dicom metadata.

marta- committed:
[misc] Minor refactoring in the Note component

marta- committed:
CARDS-1623: Add basic support for DICOM image storage
Refactoring:
- simpler incorporation of the note value that is being set programatically
- renamed `externalNote` to the more conventional `value`

marta- committed:
CARDS-1623: Add basic support for DICOM image storage
Added propTypes and defaultProps to the Note component

marta- committed:
CARDS-1623: Add basic support for DICOM image storage
Clear the contents of the note when deleting an uploaded dicom file

marta- committed:
CARDS-1623: Add basic support for DICOM image storage
Ask the user to confirm before deleting a dicom file, since now deletion involves clearing the contents of the Notes

IntegralProgrammer committed:
CARDS-1623: Add basic support for DICOM image storage

Beginning support for DICOM image preview

marta- committed:
CARDS-1623: Add basic support for DICOM image storage
Save the image extracted from the dicom file into the `image` property of the answer node

marta- committed:
CARDS-1623: Add basic support for DICOM image storage
Small refactoring: pull the local dicom processing code in a separate method

marta- committed:
CARDS-1623: Add basic support for DICOM image storage
Make the FileQuestion component customizable so question components corresponding to specific file types can leverage it

marta- committed:
CARDS-1623: Add basic support for DICOM image storage
DicomQuestion now renders a customized FileQuestion

marta- committed:
CARDS-1623: Add basic support for DICOM image storage
* Better ordering of the parameters in previewRenderer
* Escape the fileURL before passing it to the previewRenderer

marta- committed:
CARDS-1623: Add basic support for DICOM image storage
Added TODOs to DicomQuestion for switching from image saved separately to loading it from the stored DICOM file

IntegralProgrammer committed:
CARDS-1623: Add basic support for DICOM image storage

Refactored getDicomTagName and getDicomTagDataFormat to eliminate repeated code

IntegralProgrammer committed:
CARDS-1623: Add basic support for DICOM image storage

Refactor, moving the DICOM pixels to PNG DataURL conversion to its own function

IntegralProgrammer committed:
CARDS-1623: Add basic support for DICOM image storage

We no longer are storing the rendered PNG image in dataURL format in a JCR answer node property. Instead we render the PNG dataURL on-the-fly from the DICOM file stored in the JCR

IntegralProgrammer committed:
CARDS-1623: Add basic support for DICOM image storage

Clean up commenting and debugging calls

marta- committed:
CARDS-1623: Add basic support for DICOM image storage
More code and comments cleanup

marta- committed:
CARDS-1623: Add basic support for DICOM image storage
Bugfix: files with special characters in their names (e.g. `#`) cannot be previewed

IntegralProgrammer committed:
CARDS-1623: Add basic support for DICOM image storage

Improve the getDicomTagValue function so that the Value Representations (VR)s: US, UL, CS, UI, DA, TM, LO, PN, SH, DS, IS, AE can be appropriately handled with a fallback to dumping the tag value as a raw hexadecimal string

Other code cleanup

IntegralProgrammer committed:
CARDS-1724: Add support for coloured DICOM images and MONO other than knee and head

Handle RGBA pixels from colored DICOM images

IntegralProgrammer committed:
CARDS-1724: Add support for coloured DICOM images and MONO other than knee and head

Upgrade cornerstone-wado-image-loader NPM package from 4.1.2 to 4.1.3

IntegralProgrammer committed:
CARDS-1724: Add support for coloured DICOM images and MONO other than knee and head

Add the commented-out import of the dynamic-import cornerstoneWADOImageLoader package. This is there so that in future work we can use the dynamic-import version of cornerstoneWADOImageLoader which supports more DICOM formats

Show an error dialog if DICOM file parsing fails

IntegralProgrammer committed:
CARDS-1724: Add support for coloured DICOM images and MONO other than knee and head

Set showAdvancedErrorHelp to false upon close of error dialog

IntegralProgrammer committed:
CARDS-1623: Add basic support for DICOM image storage

Move the DICOM test Questionnaire out of the page-start-header module and into test-resources

IntegralProgrammer committed:
CARDS-1623: Add basic support for DICOM image storage

Remove commented-out import of dynamic loading version of cornerstoneWADOImageLoader

IntegralProgrammer committed:
CARDS-1724: Add support for coloured DICOM images and MONO other than knee and head

Refactor DicomQuestion.jsx to remove repeated code.

Do not attempt to preview DICOM images with TransferSyntaxUID values that are known to not work with the version of Cornerstone that we are using.

marta- reviewed (changes requested):
I would move dicomDataDictionary outside of `questionnaire/`, in a new directory`modules/data-entry/src/main/frontend/src/dicom/`.

marta- committed:
CARDS-1623: Add basic support for DICOM image storage
Moved dicomDataDictionary to a separate directory

marta- committed:
CARDS-1623: Add basic support for DICOM image storage
Updated DicomQuestion.jsx to Material UI 5+.

marta- committed:
CARDS-1623: Add basic support for DICOM image storage
Usability improvements to the Error dialog displayed when parsing the dicom fails

marta- commented:
Rebased on the latest `dev`, updated the new component to Material UI 5 in f3ffb1f, improved the error dialog in d0f262c.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/FileQuestion.jsx:
```diff
@@ -282,7 +288,12 @@ function FileQuestion(props) {
-    return <a href={fixFileURL(hrefs[idx], label)} target="_blank" rel="noopener" download>{label}</a>;
+    return (
+      <div>
+        <Link href={fixFileURL(hrefs[idx], label)} target="_blank" rel="noopener" download>{label}</Link>
```
This is needed in now in mui5 to keep links looking consistent:
```suggestion
        <Link href={fixFileURL(hrefs[idx], label)} target="_blank" rel="noopener" download underline="hover">{label}</Link>
```

marta- replied:
Done in bb8e3ee.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/FileQuestion.jsx:
```diff
@@ -282,7 +288,12 @@ function FileQuestion(props) {
-    return <a href={fixFileURL(hrefs[idx], label)} target="_blank" rel="noopener" download>{label}</a>;
+    return (
+      <div>
+        <Link href={fixFileURL(hrefs[idx], label)} target="_blank" rel="noopener" download>{label}</Link>
```
This is needed in now in mui5 to keep links looking consistent:
```suggestion
        <Link href={fixFileURL(hrefs[idx], label)} target="_blank" rel="noopener" download underline="hover">{label}</Link>
```

marta- committed:
CARDS-1623: Add basic support for DICOM image storage

Adjust FileQuestion UI changes after mui5 upgrade.

marta- committed:
CARDS-1623: Add basic support for DICOM image storage
Further ui improvements to the Error dialog displayed when parsing the dicom fails - display the suggested command as formatted text

sashaandjic reviewed (approved):
tested

![image](https://user-images.githubusercontent.com/23243244/172913695-94ece23a-8e4e-47f4-9408-53fd1bce35d1.png)

<img width="382" alt="image" src="https://user-images.githubusercontent.com/23243244/172913624-5747ce3c-2635-4ec7-8687-ad75ee2ee4fa.png">

sashaandjic commented:
QUESTION: Can we offer to do the conversion for the user by adding a button instead of instructions?

marta- committed:
CARDS-1623: Add basic support for DICOM image storage
Adjust the size and alignment of the delete button for both FileQuestion and DicomQuestion

marta- commented:
> QUESTION: Can we offer to do the conversion for the user by adding a button instead of giving them instructions?

It's not completely impossible, but not something we should be doing. We're not even sure that's the problem, we're just assuming they're uploading a real dcom instead of some garbage.

marta- commented:
Made two small UI adjustments in the last commits:
* Error dialog: formatted the suggested command
* File list: made the Delete button smaller and aligned it better

![image](https://user-images.githubusercontent.com/651980/172916702-157f65ae-1633-4aed-b9f3-f13c5ba601d5.png)

![image](https://user-images.githubusercontent.com/651980/172916888-c8c18825-d150-49fa-a33e-b4f983b0d939.png)

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/FileQuestion.jsx:
```diff
@@ -255,6 +260,7 @@ function FileQuestion(props) {
 
   // Delete an answer by its index
   let deletePath = (index) => {
+    onDelete && onDelete(index);
```
In the future, we could just use the optional chaining syntax `onDelete?.(index)` for checks like this.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/labels/DicomLabelProcessor.java:
```diff
@@ -0,0 +1,51 @@
+import io.uhndata.cards.serialize.spi.ResourceJsonProcessor;
+
+/**
+ * Gets the file name of the file question answer.
```
```suggestion
 * Gets the file name of a DICOM question answer.
```

marta- replied:
Updated in ecf87ba.

sdumitriu commented on modules/data-entry/src/main/resources/SLING-INF/nodetypes/dataentry.cnd:
```diff
@@ -689,6 +689,21 @@
+//-----------------------------------------------------------------------------
+// DICOM resource answers
+[cards:DicomAnswer] > cards:Answer
+  // The value holds the YAML encoded DICOM metadata
```
Does it? I thought we store the path to the file in the value, and the metadata in the notes.

marta- replied:
Fixed in ecf87ba.

marta- committed:
CARDS-1623: Add basic support for DICOM image storage
Addressing review comments: fixed copy-pasted code comments

sashaandjic commented:
re-tested 
![image](https://user-images.githubusercontent.com/23243244/172960828-e97c483a-27fd-4cce-8e8a-e81e43e25eef.png)
